src/hotspot/share/memory/archiveUtils.cpp

```diff
@@ -297,3 +301,21 @@ void ReadClosure::do_region(u_char* start, size_t size) {
     size -= sizeof(intptr_t);
   }
 }
+
+void ArchiveUtils::log_to_classlist(BootstrapInfo* bootstrap_specifier, TRAPS) {
+  if (DumpLoadedClassList != NULL && classlist_file->is_open()) {
+    if (SystemDictionaryShared::is_supported_invokedynamic(bootstrap_specifier)) {
+      ResourceMark rm(THREAD);
+      const constantPoolHandle& pool = bootstrap_specifier->pool();
+      int pool_index = bootstrap_specifier->bss_index();
+      classlist_file->print("%s %s", LAMBDA_PROXY_TAG, pool->pool_holder()->name()->as_C_string());
+      CDSIndyInfo cii;
+      ClassListParser::populate_cds_indy_info(pool, pool_index, &cii, THREAD);
+      GrowableArray<const char*>* indy_items = cii.items();
+      for (int i = 0; i < indy_items->length(); i++) {
+        classlist_file->print(" %s", indy_items->at(i));
+      }
+      classlist_file->print_cr("");
+    }
+  }
+}
```

Comment:
I think if two threads try call ArchiveUtils::log_to_classlist at the same time, the output may be interleaved.

classlist_file is a fileStream, which uses fwrite to write to the file. In theory, if you write the whole line at once, the output should be thread safe (at least on POSIX and Windows). But in your case, you would need to first get the whole line into a buffer, and then write it out at once.

I think it would be safer, and more convenient, to use a lock to ensure thready safety. Maybe we can convert all uses of classlist_file->print() to something like

```
class ClassListWriter {
    static fileStream* _classlist_file;
    MutexLocker locker;
public:
   outputStream* stream() {
       return _classlist_file;
   }
   static bool is_enabled() {
     return _classlist_file != NULL && _classlist_file->is_open();
   }
   ClassListWriter() :  locker(Thread::current(), ClassListFile_lock) {}

   static void init() {
     // classlist_file init code from ostrea.cpp
   }
};

// WAS if (DumpLoadedClassList != NULL && classlist_file->is_open()) {
if (ClassListWriter::is_enabled()) {
  ClassListWriter w;
  w->stream()->print("aaaa");
  w->stream()->print("bbbb");
  w->stream()->cr();
}
```